File: backend/app/services/url_parser.py
"""电商链接解析器。

从商品 URL 中提取平台名称和商品 ID。
"""

import re
from dataclasses import dataclass

# ---------------------------------------------------------------------------
# 平台标识常量
# ---------------------------------------------------------------------------
PLATFORM_JD = "京东"
PLATFORM_TMALL = "天猫"
PLATFORM_TAOBAO = "淘宝"
PLATFORM_PINDUODUO = "拼多多"

# 暂不支持的平台列表
UNSUPPORTED_PLATFORMS = {PLATFORM_PINDUODUO}

# ---------------------------------------------------------------------------
# URL 解析规则
# ---------------------------------------------------------------------------
# 每条规则：(正则, 平台, 商品ID所在 group 名/索引)
_RULES: list[tuple[re.Pattern, str, str]] = [
    # 京东
    (re.compile(r"item\.jd\.com/(?P<id>\d+)"), PLATFORM_JD, "id"),
    # 天猫
    (re.compile(r"detail\.tmall\.com/.*[?&]id=(?P<id>\d+)"), PLATFORM_TMALL, "id"),
    # 淘宝
    (re.compile(r"item\.taobao\.com/.*[?&]id=(?P<id>\d+)"), PLATFORM_TAOBAO, "id"),
    # 拼多多
    (
        re.compile(r"(?:yangkeduo|pinduoduo)\.com/.*goods_id=(?P<id>\d+)"),
        PLATFORM_PINDUODUO,
        "id",
    ),
    (
        re.compile(r"(?:yangkeduo|pinduoduo)\.com/.*goods[./](?P<id>\d+)"),
        PLATFORM_PINDUODUO,
        "id",
    ),
]


@dataclass
class ParsedUrl:
    """URL 解析结果。"""

    platform: str
    product_id: str
    is_supported: bool


def parse_product_url(url: str) -> ParsedUrl | None:
    """解析商品链接，提取平台和商品 ID。

    返回 None 表示无法识别的链接格式。
    """
    cleaned = url.strip()

    for pattern, platform, group in _RULES:
        match = pattern.search(cleaned)
        if match:
            product_id = match.group(group)
            return ParsedUrl(
                platform=platform,
                product_id=product_id,
                is_supported=platform not in UNSUPPORTED_PLATFORMS,
            )

    return None

File: apps/backend/src/justone/url-parser.ts
/** URL 解析器 — 从商品链接提取平台和商品 ID。移植自 Python 版 url_parser.py。 */

export const PLATFORM_JD = '京东';
export const PLATFORM_TMALL = '天猫';
export const PLATFORM_TAOBAO = '淘宝';
export const PLATFORM_PINDUODUO = '拼多多';

export const UNSUPPORTED_PLATFORMS = new Set([PLATFORM_PINDUODUO]);

interface ParsedUrl {
  platform: string;
  productId: string;
  isSupported: boolean;
}

const RULES: Array<{ pattern: RegExp; platform: string; group: string }> = [
  { pattern: /item\.jd\.com\/(?<id>\d+)/, platform: PLATFORM_JD, group: 'id' },
  { pattern: /detail\.tmall\.com\/.*[?&]id=(?<id>\d+)/, platform: PLATFORM_TMALL, group: 'id' },
  { pattern: /item\.taobao\.com\/.*[?&]id=(?<id>\d+)/, platform: PLATFORM_TAOBAO, group: 'id' },
  { pattern: /(?:yangkeduo|pinduoduo)\.com\/.*goods_id=(?<id>\d+)/, platform: PLATFORM_PINDUODUO, group: 'id' },
  { pattern: /(?:yangkeduo|pinduoduo)\.com\/.*goods[./](?<id>\d+)/, platform: PLATFORM_PINDUODUO, group: 'id' },
];

export function parseProductUrl(url: string): ParsedUrl | null {
  const cleaned = url.trim();
  for (const rule of RULES) {
    const match = rule.pattern.exec(cleaned);
    if (match) {
      const productId = match.groups?.[rule.group] || match[1];
      return {
        platform: rule.platform,
        productId,
        isSupported: !UNSUPPORTED_PLATFORMS.has(rule.platform),
      };
    }
  }
  return null;
}
